feat(colorslider): S2 migration

cdransf opened the pull request:
## Description

`CSS-1028`

# S2 colorslider migration

This migrates the `colorslider` component to S2. Custom properties have been remapped per the design spec.

| Before                    | After                      |
| ------------------------- | -------------------------- |
| `--spectrum-gray-900-rgb` | `--spectrum-gray-1000-rgb` |


### Validation steps

1. Open the URL for the PR.
2. Navigate to the `colorslider` component and verify no regressions have occurred.

### Regression testing

Validate:

1. The documentation pages for at least two other components are still loading, including:

- [ ] The pages render correctly, are accessible, and are responsive.

2. If components have been modified, VRTs have been run on this branch:

- [ ] VRTs have been run and looked at.
- [ ] Any VRT changes have been accepted (by reviewer and/or PR author), or there are no changes.

## Screenshots

<!-- If applicable, add screenshots to show what you changed -->

## To-do list

<!-- Put an "x" to indicate you've done each of the following. Add/remove additional tasks, as needed. -->

- [x] I have read the [contribution guidelines](/.github/CONTRIBUTING.md).
- [ ] I have updated relevant storybook stories and templates.
- [ ] I have tested these changes in Windows High Contrast mode.
- [ ] If my change impacts **other components**, I have tested to make sure they don't break.
- [ ] If my change impacts **documentation**, I have updated the documentation accordingly.
- [x] ✨ This pull request is ready to merge. ✨

cdransf commented:
Per design we can ignore the spacing tokens as they're not required for the implementation (alignment of the handle within the color slide is handled by this rule:

```css
.spectrum-ColorSlider-handle {
  inset-inline-start: 0;
  inset-block-start: 50%;
}
```

rise-erpelding reviewed (changes requested):
Migration-wise, this looks fine, since not a lot needed to change here! I have a few other recommendations though:

**CSS**
I have a few recommendations for the CSS that would make this file more consistent with some of the other migration work that's been done:
- Generally, we've been moving forced-colors media query blocks to the bottom in recent migrations (see #3388, #2743, or #2818 for instance), and I see that Color Slider's is still further up in the CSS file, that could probably be moved down.
- We have this line:
    ```css
    block-size: var(--mod-color-slider-control-track-width, var(--spectrum-color-control-track-width));
    ```
    I'm really not sure why that `color-control-track-width` token has that name, but it seems like a good idea to create an alias custom property for this, especially given that its token name doesn't match its mod name, something like this, we can even take the word "control" out and change the mod name completely, up to you:
    ```css
     /* Up at the topmost .spectrum-ColorSlider block */
     --spectrum-color-slider-control-track-width: var(--spectrum-color-control-track-width);

    /* further down in the code, same place it is now */
    block-size: var(--mod-color-slider-control-track-width, var(--spectrum-color-slider-control-track-width));
    ```
- In that same vein, we could also alias the global token for the disabled background color that we have here, similar to how it's done in [Color Wheel](https://github.com/adobe/spectrum-css/blob/spectrum-two/components/colorwheel/index.css#L18), [Picker](https://github.com/adobe/spectrum-css/blob/spectrum-two/components/picker/index.css#L100), or [Search](https://github.com/adobe/spectrum-css/blob/spectrum-two/components/search/index.css#L48), for instance:
    ```css
    --spectrum-color-slider-border-color-local: var(--highcontrast-color-slider-border-color-disabled, var(--mod-color-slider-border-color-disabled, var(--spectrum-disabled-background-color)));
    background: var(--highcontrast-color-slider-background-color-disabled, var(--mod-color-slider-background-color-disabled, var(--spectrum-disabled-background-color)));
    ```
- There's also an unused token for the spacing between the field label and the color slider, but this isn't new. I don't know if there's some specific reason we're not adding the label to this implementation (Slider has a label that we display), but it feels like something we ought to be doing. Feel free to get more opinions on this one.
<img width="720" alt="image" src="https://github.com/user-attachments/assets/62ed49d9-8017-4b17-a2f5-e936e3cf526c" />

**Storybook**
Otherwise, this component could probably benefit from a few Storybook enhancements similar to what we added for several of our other components back when we were doing docs to storybook migrations (this component got migrated early on, before we decided that we would be doing that kind of thing):
- Being able to change the colors in the slider from the Storybook controls or add an image (Swatch on main added this with the docs to storybook migration https://github.com/adobe/spectrum-css/pull/2925, so we might be able to copy that over)
- A control to be able to view the vertical variant from the default story in Storybook. In this branch, all the variants are showing, but on `main`, on the default variant is visible and the others are used just for display purposes on the Docs page, that's probably where we'd like to end up one day when `spectrum-two` has some of those docs changes from main.

Last thing! I can't use the keyboard to focus on the slider handle (I can in `main`, but not on this branch or `spectrum-two`), this may not be a big deal, but I'm curious about why it's happening.

cdransf commented:
> Migration-wise, this looks fine, since not a lot needed to change here! I have a few other recommendations though:
> 
> **CSS** I have a few recommendations for the CSS that would make this file more consistent with some of the other migration work that's been done:
> 
> * Generally, we've been moving forced-colors media query blocks to the bottom in recent migrations (see [feat(colorarea): S2 migration #3388](https://github.com/adobe/spectrum-css/pull/3388), [feat(tooltip): S2 migration #2743](https://github.com/adobe/spectrum-css/pull/2743), or [feat(statuslight): s2 migration #2818](https://github.com/adobe/spectrum-css/pull/2818) for instance), and I see that Color Slider's is still further up in the CSS file, that could probably be moved down.

Alrighty! `forced-colors` has been moved to the bottom of the file.

> * We have this line:
>   ```css
>   block-size: var(--mod-color-slider-control-track-width, var(--spectrum-color-control-track-width));
>   ```
>     
>   I'm really not sure why that `color-control-track-width` token has that name, but it seems like a good idea to create an alias custom property for this, especially given that its token name doesn't match its mod name, something like this, we can even take the word "control" out and change the mod name completely, up to you:
>   ```css
>    /* Up at the topmost .spectrum-ColorSlider block */
>    --spectrum-color-slider-control-track-width: var(--spectrum-color-control-track-width);
>   
>   /* further down in the code, same place it is now */
>   block-size: var(--mod-color-slider-control-track-width, var(--spectrum-color-slider-control-track-width));
>   ```

`--spectrum-color-control-track-width` has been aliased to `--spectrum-color-slider-control-track-width`.

> * In that same vein, we could also alias the global token for the disabled background color that we have here, similar to how it's done in [Color Wheel](https://github.com/adobe/spectrum-css/blob/spectrum-two/components/colorwheel/index.css#L18), [Picker](https://github.com/adobe/spectrum-css/blob/spectrum-two/components/picker/index.css#L100), or [Search](https://github.com/adobe/spectrum-css/blob/spectrum-two/components/search/index.css#L48), for instance:
>   ```css
>   --spectrum-color-slider-border-color-local: var(--highcontrast-color-slider-border-color-disabled, var(--mod-color-slider-border-color-disabled, var(--spectrum-disabled-background-color)));
>   background: var(--highcontrast-color-slider-background-color-disabled, var(--mod-color-slider-background-color-disabled, var(--spectrum-disabled-background-color)));
>   ```

`--spectrum-disabled-background-color` has been aliased to `--spectrum-colorslider-fill-color-disabled`.

> * There's also an unused token for the spacing between the field label and the color slider, but this isn't new. I don't know if there's some specific reason we're not adding the label to this implementation (Slider has a label that we display), but it feels like something we ought to be doing. Feel free to get more opinions on this one.

@jawinn and I discussed the label visible in this component and (apparently) SWC doesn't have the label as part of the component and the consensus was that we didn't need to include it at this point.

> **Storybook** Otherwise, this component could probably benefit from a few Storybook enhancements similar to what we added for several of our other components back when we were doing docs to storybook migrations (this component got migrated early on, before we decided that we would be doing that kind of thing):
> 
> * Being able to change the colors in the slider from the Storybook controls or add an image (Swatch on main added this with the docs to storybook migration [docs(swatch,swatchgroup,table,tabs): docs migrations to storybook #2925](https://github.com/adobe/spectrum-css/pull/2925), so we might be able to copy that over)
> * A control to be able to view the vertical variant from the default story in Storybook. In this branch, all the variants are showing, but on `main`, on the default variant is visible and the others are used just for display purposes on the Docs page, that's probably where we'd like to end up one day when `spectrum-two` has some of those docs changes from main.

I added a `colors` control to the story for the component that defaults to the previously hardcoded linear gradient and a `vertical` boolean control that defaults to false and allows the orientation to be changed.

> Last thing! I can't use the keyboard to focus on the slider handle (I can in `main`, but not on this branch or `spectrum-two`), this may not be a big deal, but I'm curious about why it's happening.

I can reproduce this issue too, but haven't been able to sort out why it's happening yet. 🤔

cdransf commented:
I am able to focus on the handle in this branch (testing in Chrome) — I can't in Safari, but I often get odd keyboard focus behavior in Safari.

<img width="1104" alt="image" src="https://github.com/user-attachments/assets/6a214ee2-d3bd-47d4-bdc0-b2c0b8283b68" />

rise-erpelding reviewed (commented):
Looking great! I love the improvements to Storybook, and for whatever reason the keyboard focus issue has resolved itself! 🎉 

One minor nitpick about the `--spectrum-colorslider-fill-color-disabled` custom prop.

cdransf commented on components/colorslider/metadata/metadata.json:
```diff
@@ -37,13 +36,15 @@
     "--spectrum-color-slider-length",
-    "--spectrum-color-slider-minimum-length"
+    "--spectrum-color-slider-minimum-length",
+    "--spectrum-colorslider-fill-color-disabled"
```
Makes sense to me! Just updated it. ✨

jawinn commented on components/colorslider/stories/colorslider.stories.js:
### Alpha 
It looks like the Alpha story is still using the `gradientStops` arg that renamed, so it's not displaying correctly:
<img width="440" alt="Screenshot 2025-01-17 at 3 53 30 PM" src="https://github.com/user-attachments/assets/a706c829-b950-47a8-a607-9cc5cf7e33b5" />
Prod:
<img width="243" alt="Screenshot 2025-01-17 at 3 58 38 PM" src="https://github.com/user-attachments/assets/f4c7da53-1e94-4b01-aaad-40a2041aff44" />

### Vertical 
The Vertical story also still needs an update for its story in our docs, and the VRT test. The gradient should be in the `to bottom` direction for these instances where we set the arg value, though we talked about not needing any complicated logic if someone is manually flipping the vertical control on the Default story:

<img width="478" alt="Screenshot 2025-01-17 at 3 53 16 PM" src="https://github.com/user-attachments/assets/d5ee5089-2cfd-45c6-a264-e83f158e49b5" />
<img width="415" alt="Screenshot 2025-01-17 at 3 54 08 PM" src="https://github.com/user-attachments/assets/49b52260-ff5c-4079-b78d-5d0ba9cd1cc8" />

cdransf replied:
Got it! I renamed and updated that arg and updated the gradient direction in the tests you highlighted. ✨

jawinn reviewed (changes requested):
I'm still seeing a few issues with the Alpha and Vertical stories being displayed on the Docs.

This'll also need some rebasing after the large rebase in `spectrum-two` that recently happened, which is causing the large number of changed files to be shown. It may be easier to cherry-pick your commits than to handle the conflicts (at least that's what I found in my other branch).

rise-erpelding reviewed (changes requested):
A couple of other details on the docs page that look like they still need some attention:

* This Alpha story doesn't look the same as in Foundations or main, it looks like the gradient in the args is using the same rainbow gradient rather than the previous one that included rgba values.
    ![image](https://github.com/user-attachments/assets/d89ad085-6dc7-48f5-ac2f-57ef2d311a38)
* The Vertical variant is improved! But it looks like it needs its handle adjusted, the color doesn't match the color next to it in the gradient and it's off-center. Previous versions showed the handle at the top which probably isn't strictly necessary, but it seems like maybe that's where the red color is coming from. Feels like a bit of a silly ask since it's not a working implementation, but would help it look the part anyway!
    ![image](https://github.com/user-attachments/assets/fa6d8bd2-4258-40af-8120-b74d9b1a78e3)

cdransf commented:
> A couple of other details on the docs page that look like they still need some attention:
> 
> * This Alpha story doesn't look the same as in Foundations or main, it looks like the gradient in the args is using the same rainbow gradient rather than the previous one that included rgba values.

Totally missed that! Thank you! That should be all fixed (updated with the value from the `redux` branch). ✨

> * The Vertical variant is improved! But it looks like it needs its handle adjusted, the color doesn't match the color next to it in the gradient and it's off-center. Previous versions showed the handle at the top which probably isn't strictly necessary, but it seems like maybe that's where the red color is coming from. Feels like a bit of a silly ask since it's not a working implementation, but would help it look the part anyway!

I adjusted the `inset-block-start` for the `.spectrum-ColorSlider-handle` when the slider is oriented vertically. The default class hasn't been updated in ~2-5 years.

```
.spectrum-ColorSlider-handle {
  inset-inline-start: 0;
  inset-block-start: 50%;
}
```

Which is why we're seeing the handle 50% of the way through the slider. I added:

```
.spectrum-ColorSlider--vertical {
  ...
  .spectrum-ColorSlider-handle {
    inset-block-start: 0;
  }
}
```

Which I'm not sure is the right path — would love some feedback on that. 😄 

<img width="673" alt="image" src="https://github.com/user-attachments/assets/11d53bfb-9a9f-44e5-9f88-b5ca66b274f1" />

rise-erpelding commented:
Hey! The docs page Alpha story looks good, I think it's still going to need an update for the testing preview:
![image](https://github.com/user-attachments/assets/c84bded3-e5d6-4964-9698-893614f9e0fb)

Re: the vertical story, it looks like maybe that `--vertical` variant is just missing the `inset-inline-start: 50%` like it had in [Foundations](https://github.com/adobe/spectrum-css/blob/s2-foundations-redux/components/colorslider/index.css#L64)?

rise-erpelding commented on components/colorslider/stories/colorslider.stories.js:
```diff
@@ -40,12 +42,19 @@ export default {
 			control: { type: "select" },
 			table: { disable: true },
 		},
+		vertical: {
+			name: "Vertical",
+			description: "The orientation of the component.",
+			type: { name: "boolean" },
+			control: "boolean",
+		},
```
Ooof, I'm still seeing that funny gradient when the vertical control is used:
![image](https://github.com/user-attachments/assets/37b01f2f-f6c3-4c43-82a3-726fdaf8de03)

I don't have ideas off the top of my head on how we handle this, since it would need to be one gradient for horizontal and another for vertical... although I'm sure we've run into a similar issue before... any thoughts?

cdransf replied:
I talked to @jawinn about this and the thinking at the time was that, since we've moved the color control to a string that the onus would then be on the user to change that direction. We could couple the two, I can investigate that a bit more and see what we'd need to do it. 😃

rise-erpelding replied:
I think that's fine. We could add a little blurb in the Color control description about the gradient direction, maybe saying that the direction would need to be changed for the vertical variant. Not sure I feel strongly about that though, so I'll leave that up to you!

rise-erpelding reviewed (changes requested):
Looks like the Alpha gradient in the tests and the handle for the vertical variant still need to be addressed, I also left another comment about some other weirdness with the vertical gradient.

cdransf commented on components/colorslider/index.css:
```diff
@@ -55,6 +55,10 @@
+
+	.spectrum-ColorSlider-handle {
+		inset-block-start: 0;
+	}
```
Just restored this! ✨

cdransf commented on components/colorslider/stories/colorslider.test.js:
```diff
@@ -14,14 +14,15 @@ export const ColorSliderGroup = Variants({
 		{
 			testHeading: "Alpha",
-			gradientStops: ["rgba(0, 0, 0, 1) 0%", "rgba(0, 0, 0, 0) 100%"],
+			color: "linear-gradient(to right, rgb(255, 0, 0) 0%, rgb(255, 255, 0) 17%, rgb(0, 255, 0) 33%, rgb(0, 255, 255) 50%, rgb(0, 0, 255) 67%, rgb(255, 0, 255) 83%, rgb(255, 0, 0) 100%)",
```
Updated this as well. ✨

rise-erpelding reviewed (approved):
Looks great, I think everything I had has been addressed now! Thanks Cory!
